streaming/src/event_server.h

```diff
@@ -0,0 +1,229 @@
+#ifndef RAY_STREAMING_EVENT_SERVER_H
+#define RAY_STREAMING_EVENT_SERVER_H
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <unordered_map>
+
+#include "channel.h"
+#include "ring_buffer.h"
+#include "util/streaming_util.h"
+
+namespace ray {
+namespace streaming {
+
+/// Data writer utilizes what's called an event-driven programming model instead
+/// loop forward polling driven that's employed in first simple version.
+/// Event driven model includes two important components: event server and event
+/// queue. In the process of data transmission, the inputer will first define
+/// the processing method of corresponding events. However, by triggering
+/// different events in actual operation, these events will be put into the event
+/// queue, and finally the event server will schedule the previously registered
+/// processing functions ordered by its priority.
+
+/// The event queue inherits from the universal ring queue. It mainly divides
+/// event into two different levels: normal event and urgent event, and the
+/// total size of the queue is the sum of them.
+template <class T>
+class EventQueue : public AbstractRingBufferImpl<T> {
+ public:
+  EventQueue(size_t size) : capacity_(size), is_started_(true) {}
+
+  virtual ~EventQueue() {
+    is_started_ = false;
+    full_cv_.notify_all();
+    empty_cv_.notify_all();
+  };
+
+  void Run() { is_started_ = true; }
+
+  void Stop() {
+    is_started_ = false;
+    empty_cv_.notify_all();
+    full_cv_.notify_all();
+  }
+
+  void Push(const T &t) {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Size() >= capacity_ && is_started_) {
+      STREAMING_LOG(WARNING) << " EventQueue is full, its size:" << Size()
+                             << " all_size:" << Size() << " capacity:" << capacity_
+                             << " buffer size:" << buffer_.size()
+                             << " urgent_buffer size:" << urgent_buffer_.size();
+      full_cv_.wait(lock);
+      STREAMING_LOG(WARNING) << "Event server is full_sleep be notified";
+    }
+    if (!is_started_) {
+      return;
+    }
+    buffer_.push(t);
+    if (1 == Size()) {
+      empty_cv_.notify_one();
+    }
+  }
+
+  void PushToUrgent(const T &t) {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
```

Comment:
this function looks very similar to the Push function above, consider merging them into one.

Reply:
Good catch.
There is only one difference that event will be added to urgent buffer not normal queue in `Push` function.
But I will try to merge them into one.